netket/experimental/driver/infidelity_sr.py

```diff
@@ -0,0 +1,497 @@
+
+        Args:
+            target_state: The target state that must be matched.
+            optimizer: The optimizer to use for the parameter updates. To perform proper
+                SR/NGD optimization this should be an instance of `optax.sgd`, but can be
+                any other optimizer if you are brave.
+            operator: The operator U.
+            variational_state: The variational state to optimize.
+            diag_shift: The diagonal regularization parameter :math:`\lambda` for the QGT/NTK.
+            proj_reg: The regularization parameter for the projection of the updates.
+                (This usually is not very important and can be left to None)
+            momentum: (SPRING, disabled by default, read above for details) a number between [0,1]
+                that specifies the damping factor of
+                the previous updates and works somewhat similarly to the beta parameter of ADAM.
+                The maximum amplification of  the step size in SPRING is
+                :math:`A(\mu)=1/\sqrt{1-μ^2}`
+                Thus the  amplification is at most a factor of :math:`A(0.9)=2.3` or
+                :math:`A(0.99)=7.1`. Values around ``momentum = 0.8`` empirically work well.
+                (Defaults to None)
+            linear_solver_fn: The linear solver function to use for the NGD solver.
+            mode: The mode used to compute the jacobian of the variational state.
+                Can be `'real'` or `'complex'`. Real can be used for real-valued wavefunctions
+                with a sign, to truncate the arbitrary phase of the wavefunction. This leads
+                to lower computational cost.
+            on_the_fly: Whether to compute the QGT or NTK using lazy evaluation methods.
+                This usually requires less memory. (Defaults to None, which will
+                automatically chose the potentially best method).
+            chunk_size_bwd: The number of rows of the NTK or of the Jacobian evaluated
+                in a single sweep.
+            use_ntk: Wheter to compute the updates using the Neural Tangent Kernel (NTK)
+                instead of the Quantum Geometric Tensor (QGT), aka switching between
+                SR and minSR. (Defaults to None, which will automatically choose the best
+                method)
+        """
+
+        if operator is not None:
+
+            def _logpsi_fun(apply_fun, variables, x, *args):
+                variables_applyfun, O = flax.core.pop(variables, "operator")
+
+                xp, mels = O.get_conn_padded(x)
+                xp = xp.reshape(-1, x.shape[-1])
+                logpsi_xp = apply_fun(variables_applyfun, xp, *args)
+                logpsi_xp = logpsi_xp.reshape(mels.shape)
+
+                return jax.scipy.special.logsumexp(
+                    logpsi_xp.astype(complex), axis=-1, b=mels
+                )
+
+            logpsi_fun = HashablePartial(_logpsi_fun, target_state._apply_fun)
+
```

Comment:
Can you put this in a file? something like https://gist.github.com/PhilipVinc/5fcc9fbba0e5d6ed6d56f0a4325e2808 
you could put in `_src/operator/hpsi_utils.py